# QuickCart Capstone — Hypothesis Testing Workflow

A fictional e-commerce company, QuickCart, wants to make several
data-driven decisions using hypothesis testing. This notebook walks
through 4 business questions, choosing and applying the right
statistical test for each, and closes with a summary report.

In [1]:
import numpy as np
from scipy import stats

## Task 1 — Website Load Time Claim

**Question:** QuickCart's engineering team claims the site loads in 2.0
seconds on average. Population std is known (0.5s). A sample of 45 page
loads this week averaged 2.15 seconds. Has load time actually changed?

**Data type:** Numeric (load time is a continuous measurement, not a category)

**Test to use:** Z-test — comparing one sample against a known claimed
value, with population standard deviation known.

**Hypotheses:**
- H0: The true average load time is 2.0 seconds (matches the claim)
- H1: The true average load time is NOT 2.0 seconds (differs from the claim)

In [2]:
sample_mean = 2.15
pop_mean = 2.0
pop_std = 0.5
n = 45
alpha = 0.05

z = (sample_mean - pop_mean) / (pop_std / np.sqrt(n))

p_value = 2 * (1 - stats.norm.cdf(abs(z)))

print("Z - statistic:", round(z, 4))
print("P - value:", round(p_value, 4))

if p_value <= alpha:
    print("reject H0")
else:
    print("fail to reject H0")

Z - statistic: 2.0125
P - value: 0.0442
reject H0


**Conclusion:** Since the p-value (0.0442) is below alpha (0.05), we
reject H0. There is statistically significant evidence that the site's
true average load time has changed from the claimed 2.0 seconds —
consistent with the sample averaging 2.15 seconds. However, this result
is close to the significance threshold (0.0442 vs 0.05), so it's worth
monitoring rather than treating as a major issue.

## Task 2 — Two Marketing Email Subject Lines

**Question:** QuickCart A/B tested two email subject lines, tracking
minutes-to-first-purchase for customers who clicked through. Population
std is unknown for this data. Does one subject line lead to faster
purchases than the other?

**Data type:** Numeric (minutes-to-first-purchase is a continuous
measurement, not a category)

**Test to use:** Two-sample t-test — comparing two independent groups
(subject line A customers vs subject line B customers), with population
standard deviation unknown.

**Hypotheses:**
- H0: There is NO difference in true average minutes-to-first-purchase
  between subject line A and subject line B (any observed difference is
  random chance)
- H1: There IS a difference in true average minutes-to-first-purchase
  between subject line A and subject line B

In [4]:
subject_a = np.array([12, 15, 9, 18, 14, 11, 16, 13, 10, 17])
subject_b = np.array([8, 11, 7, 13, 9, 10, 12, 8, 9, 11])
alpha = 0.05

t_statistic, p_value = stats.ttest_ind(subject_a, subject_b)

print("T - Statistic:", round(t_statistic, 4))
print("P - value:", round(p_value, 4))

if p_value <= alpha:
    print("reject H0")
else:
    print("fail to reject H0")

T - Statistic: 3.2577
P - value: 0.0044
reject H0


**Conclusion:** Since the p-value (0.0044) is well below alpha (0.05),
we reject H0. There is strong statistical evidence of a real difference
in average minutes-to-first-purchase between the two subject lines.
Looking at the raw data, Subject Line B has consistently lower
minutes-to-first-purchase (customers buy faster) than Subject Line A --
QuickCart should consider adopting Subject Line B going forward.

## Task 3 — Customer Support Ticket Category Distribution

**Question:** QuickCart's support team historically assumes tickets
split evenly across 4 categories: Shipping, Billing, Product, Account.
This month, out of 200 tickets, has the actual distribution shifted
from that even split?

**Data type:** Categorical (ticket counts within 4 categories)

**Test to use:** Chi-square goodness-of-fit test — comparing ONE
categorical variable's observed distribution against an expected
(evenly split) distribution.

**Hypotheses:**
- H0: Support tickets ARE evenly split across the 4 categories
  (Shipping, Billing, Product, Account) — matches the historical
  assumption
- H1: Support tickets are NOT evenly split across the 4 categories —
  the distribution has shifted

In [13]:
observed_tickets = np.array([70, 35, 55, 40])
n_total_tickets = 200
n_categories = 4
alpha = 0.05

expected = np.array([n_total_tickets / n_categories] * n_categories)

test_stat, p_value = stats.chisquare(f_obs=observed_tickets, f_exp=expected)

print("test statistic:", round(test_stat, 4))
print("p-value:", round(p_value, 4))

if p_value <= alpha:
    print("reject H0")
else:
    print("fail to reject H0")

test statistic: 15.0
p-value: 0.0018
reject H0


**Conclusion:** Since the p-value (0.0018) is well below alpha (0.05),
we reject H0. There is strong statistical evidence that support tickets
are NOT evenly split across the 4 categories this month. Looking at the
observed counts (Shipping: 70, Billing: 35, Product: 55, Account: 40),
Shipping tickets are notably higher than the expected even split of 50
each -- QuickCart may want to investigate what's driving the increase in
shipping-related issues.

## Task 4 — Shipping Times Across 4 Warehouse Regions

**Question:** QuickCart wants to know if average shipping time (in days)
differs across its 4 warehouse regions: North, South, East, West.

## Task 4 — Shipping Times Across 4 Warehouse Regions

**Question:** QuickCart wants to know if average shipping time (in days)
differs across its 4 warehouse regions: North, South, East, West.

**Data type:** Numeric (shipping time in days is a continuous measurement)

**Test to use:** One-way ANOVA — comparing the means of 4 independent
groups (regions) simultaneously.

**Hypotheses:**
- H0: The true average (mean) shipping times are EQUAL across all 4
  regions (North = South = East = West)
- H1: At least ONE region's true average (mean) shipping time is
  DIFFERENT from the others

In [14]:
region_north = np.array([3.1, 2.8, 3.4, 3.0, 2.9, 3.2])
region_south = np.array([4.2, 3.9, 4.5, 4.1, 4.0, 4.3])
region_east = np.array([3.0, 3.3, 2.9, 3.1, 3.2, 2.8])
region_west = np.array([3.2, 2.9, 3.1, 3.4, 3.0, 3.3])
alpha = 0.05

f_statistic, p_value = stats.f_oneway(region_north, region_south, region_east, region_west)

print("f-statistic:", round(f_statistic, 4))
print("p-value:", round(p_value, 4))

if p_value <= alpha:
    print("reject H0")
else:
    print("fail to reject H0")

f-statistic: 42.9524
p-value: 0.0
reject H0


**Conclusion:** Since the p-value (~0.0) is far below alpha (0.05), we
reject H0. There is strong statistical evidence that average shipping
times differ significantly across the 4 warehouse regions. Looking at
the raw data, the South region stands out with noticeably longer
shipping times (3.9-4.5 days) compared to North, East, and West (all
roughly 2.8-3.4 days) -- QuickCart should investigate what's causing
delays specifically in the South region.

# QuickCart — Hypothesis Testing Summary Report

## 1. Website Load Time
- **Question:** Does the site load in the claimed 2.0 seconds on average?
- **Test:** Z-test (one sample vs known value, population σ known)
- **Result:** z=2.0125, p=0.0442 → reject H0
- **Finding:** There is statistically significant evidence that the site's
true average load time has changed from the claimed 2.0 seconds —
consistent with the sample averaging 2.15 seconds.

## 2. Email Subject Line A/B Test
- **Question:** A/B tested two email subject lines and tracked minutes-to-first-purchase for customers who clicked through
- **Test:** two-sample t-test
- **Result:** T - Statistic: 3.2577, P - value: 0.0044, reject H0
- **Finding:** There is strong statistical evidence of a real difference
in average minutes-to-first-purchase between the two subject lines.

## 3. Support Ticket Distribution
- **Question:** support team historically assumes tickets split evenly across 4 categories
- **Test:** Chi-square goodnees-of-fit-test
- **Result:** test statistic: 15.0, p-value: 0.0018, reject H0
- **Finding:** There is strong statistical evidence that support tickets
are NOT evenly split across the 4 categories this month.

## 4. Shipping Times Across Regions
- **Question:** wants to know if average shipping time (in days) differs across its 4 warehouse regions
- **Test:** ANOVA
- **Result:** f-statistic: 42.9524, p-value: 0.0, reject H0
- **Finding:** There is strong statistical evidence that average shipping
times differ significantly across the 4 warehouse regions.

## Overall Recommendations
The most urgent priorities are the South region's shipping delays
(Task 4) and the spike in Shipping-related support tickets (Task 3) —
both show strong statistical evidence (p<0.01) and likely indicate a
real operational problem worth investigating together, since slower
shipping in the South may be directly driving the extra shipping
complaints. Adopting Subject Line B (Task 2) is a quick, low-risk win
that can be implemented immediately. The load time increase (Task 1) is
a borderline result (p=0.0442) and should be monitored, but doesn't
require urgent action on its own.